components/camel-nats/src/main/java/org/apache/camel/component/nats/NatsConsumer.java

```diff
@@ -123,37 +131,88 @@ class NatsConsumingTask implements Runnable {
-                        NatsConsumer.this.dispatcher.unsubscribe(
-                                NatsConsumer.this.getEndpoint().getConfiguration().getTopic(),
-                                Integer.parseInt(NatsConsumer.this.getEndpoint().getConfiguration().getMaxMessages()));
-                    }
-                    if (NatsConsumer.this.dispatcher.isActive()) {
-                        NatsConsumer.this.setActive(true);
-                    }
+                NatsConfiguration config = getEndpoint().getConfiguration();
+                String topic = config.getTopic();
+                String queueName = config.getQueueName();
+                String maxMessagesStr = config.getMaxMessages();
+                Integer maxMessages = null;
+                if (ObjectHelper.isNotEmpty(maxMessagesStr)) {
+                    maxMessages = Integer.parseInt(maxMessagesStr);
+                }
+
+                if (config.isJetstreamEnabled() && connection.getServerInfo().isJetStreamAvailable()) {
+                    setupJetStreamConsumer(topic, queueName);
                 } else {
-                    NatsConsumer.this.dispatcher = NatsConsumer.this.dispatcher
-                            .subscribe(NatsConsumer.this.getEndpoint().getConfiguration().getTopic());
-                    if (ObjectHelper.isNotEmpty(NatsConsumer.this.getEndpoint().getConfiguration().getMaxMessages())) {
-                        NatsConsumer.this.dispatcher.unsubscribe(
-                                NatsConsumer.this.getEndpoint().getConfiguration().getTopic(),
-                                Integer.parseInt(NatsConsumer.this.getEndpoint().getConfiguration().getMaxMessages()));
-                    }
-                    if (NatsConsumer.this.dispatcher.isActive()) {
-                        NatsConsumer.this.setActive(true);
-                    }
+                    setupStandardNatsConsumer(topic, queueName, maxMessages);
                 }
             } catch (final Exception e) {
                 NatsConsumer.this.getExceptionHandler().handleException("Error during processing", e);
             }
 
         }
 
+        private void setupJetStreamConsumer(String topic, String queueName) throws IOException, JetStreamApiException {
+            String streamName = this.configuration.getJetstreamName();
+            String consumerName
+                    = ObjectHelper.isNotEmpty(queueName) ? queueName : "consumer-" + System.currentTimeMillis(); // Generate a default consumer name if queueName is not provided
+            LOG.debug("Setting up JetStream PUSH consumer for stream: '{}', durable: '{}', topic: {} ", streamName,
+                    consumerName, this.configuration.getTopic());
+
+            JetStreamManagement jsm = connection.jetStreamManagement();
+            try {
+                // Try to get the stream, create it if it doesn't exist
```

Comment:
Sorry for my late comment on something being already merged, but instead of trying to make use of a (potentially) thrown Exception as the flow control for the logic:

https://ahdak.github.io/blog/effective-java-part-9/

Can't we make use of another API to find it out beforehand if stream available, and then create if required? Maybe using?

[JetStreamManagement#getStreams(String subjectFilter)](https://github.com/nats-io/nats.java/blob/b1b12b4194f479dae3cfed4e130e64a3ebcbee45/src/main/java/io/nats/client/JetStreamManagement.java#L250)

Reply:
I will create another PR with modified code.

Reply:
https://github.com/apache/camel/pull/18274